ecs-agent/api/ecs/client/ecs_client.go

```diff
@@ -0,0 +1,802 @@
+	attribs := attributes
+	for _, attribute := range attribs {
+		attributeMap[aws.StringValue(attribute.Name)] = aws.StringValue(attribute.Value)
+	}
+	return attributeMap
+}
+
+func findMissingAttributes(expectedAttributes, actualAttributes map[string]string) ([]string, error) {
+	missingAttributes := make([]string, 0)
+	var err error
+	for key, val := range expectedAttributes {
+		if actualAttributes[key] != val {
+			missingAttributes = append(missingAttributes, key)
+		} else {
+			logger.Trace("Response contained expected value for attribute", logger.Fields{
+				"key": key,
+			})
+		}
+	}
+	if len(missingAttributes) > 0 {
+		err = apierrors.NewAttributeError("Attribute validation failed")
+	}
+	return missingAttributes, err
+}
+
+func (client *ecsClient) getResources() ([]*ecsmodel.Resource, error) {
+	// Below are micro-optimizations - the pointers to integerStr and stringSetStr are used multiple times below.
+	integerStr := "INTEGER"
+	stringSetStr := "STRINGSET"
+
+	cpu, mem := getCpuAndMemory()
+	remainingMem := mem - int64(client.configAccessor.ReservedMemory())
+	logger.Info("Remaining memory", logger.Fields{
+		"remainingMemory": remainingMem,
+	})
+	if remainingMem < 0 {
+		return nil, fmt.Errorf(
+			"api register-container-instance: reserved memory is higher than available memory on the host, "+
+				"total memory: %d, reserved: %d", mem, client.configAccessor.ReservedMemory())
+	}
+
+	cpuResource := ecsmodel.Resource{
+		Name:         aws.String("CPU"),
+		Type:         &integerStr,
+		IntegerValue: &cpu,
+	}
+	memResource := ecsmodel.Resource{
+		Name:         aws.String("MEMORY"),
+		Type:         &integerStr,
+		IntegerValue: &remainingMem,
+	}
+	portResource := ecsmodel.Resource{
+		Name:           aws.String("PORTS"),
+		Type:           &stringSetStr,
+		StringSetValue: utils.Uint16SliceToStringSlice(client.configAccessor.ReservedPorts()),
+	}
+	udpPortResource := ecsmodel.Resource{
+		Name:           aws.String("PORTS_UDP"),
+		Type:           &stringSetStr,
+		StringSetValue: utils.Uint16SliceToStringSlice(client.configAccessor.ReservedPortsUDP()),
+	}
+
+	return []*ecsmodel.Resource{&cpuResource, &memResource, &portResource, &udpPortResource}, nil
+}
+
+// GetHostResources calling getHostResources to get a list of CPU, MEMORY, PORTS and PORTS_UPD resources
+// and return a resourceMap that map the resource name to each resource
+func (client *ecsClient) GetHostResources() (map[string]*ecsmodel.Resource, error) {
+	resources, err := client.getResources()
+	if err != nil {
+		return nil, err
+	}
+	resourceMap := make(map[string]*ecsmodel.Resource)
+	for _, resource := range resources {
+		if *resource.Name == "PORTS" {
+			// Except for RCI, TCP Ports are named as PORTS_TCP in Agent for Host Resources purpose.
+			resource.Name = aws.String("PORTS_TCP")
+		}
+		resourceMap[*resource.Name] = resource
+	}
+	return resourceMap, nil
+}
+
+func getCpuAndMemory() (int64, int64) {
+	memInfo, err := system.ReadMemInfo()
```

Comment:
I believe `system.ReadMemInfo()` is deprecated in favor of `meminfo.Read()`

source: https://pkg.go.dev/github.com/moby/moby/pkg/system#ReadMemInfo

Reply:
Thanks for calling out, will update this.